Implement Thumbnail Detection Fix for Kobo 5.6 variant

TomRomeo opened the pull request:
This pull request aims to account for the different thumbnail handling on Tolino devices with DbVersion 191 and above. For those newer versions, the dot before the file extension has to be preserved.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/NiLuJe/kfmon/15)
<!-- Reviewable:end -->

TomRomeo committed:
Implement Fix for Kobo 5.6 variant

TomRomeo committed:
fix: move sqlite3_finalize down to destroy the statement later

beedaddy commented:
One limitation could be that if someone has installed KFMon before FW version 5.6, the file name is still without a dot despite (meanwhile) DbVersion >= 191. I'm not sure if we should handle this? @NiLuJe?
(On the other hand, deleting and re-importing the .png files is enough to fix this.)

beedaddy commented on kfmon.c:
```diff
@@ -1481,20 +1479,50 @@ static bool
+					// Fallback to the v5 variant
+					replace_invalid_chars(converted_book_path);
+				}
+				sqlite3_finalize(stmt);
```
As I mentioned, I'm not so good at C. But is sqlite3_finalize correct here? Isn't this done a few lines further down anyway?

TomRomeo replied:
You are right, I didn't see that there was a sqlite3_finalize statement below anyways. I will remove it here.

NiLuJe commented:
> I'm not sure if we should handle this?

I was thinking of just brute-forcing both checks, much like how the v5 check is done only if the v4 fails ;).

(Which was a quick fix, mind you, ideally *that* one could be branched off properly based on https://github.com/NiLuJe/kfmon/blob/2ba6068c8c755971196773a2c7dbe33be6246b42/kfmon.c#L2921-L2924; but here, given that old munging on new db is going to be *very* common, I don't really see a way around checking both ways).

NiLuJe commented:
> I was thinking of just brute-forcing both checks, much like how the v5 check is done only if the v4 fails ;).

Which would solve your statement connundrum, because no more SQL; and allow to preserve the logging, because I *really* want to log which checks failed, and not a generic message ;).

TomRomeo committed:
fix: brute-force / chain thumbnail variants instead of DbVersion check

TomRomeo commented:
> > I'm not sure if we should handle this?
> 
> I was thinking of just brute-forcing both checks, much like how the v5 check is done only if the v4 fails ;).

Yeah, we probably can't work around that. I have adapted the code, removed the SQL statement and refactored the cases to be both checked sequentially. However this has led to quite some duplicate code. Do you think it might be worth to refactor this a little bit here or is it fine like that?

beedaddy commented:
Looks good to me.

NiLuJe commented:
Yeah, I might refactor that at some point, but this'll have to do for now ;).

NiLuJe commented:
Thanks!

NiLuJe commented:
Okay, I pushed a refactor to try to limit code duplication (and, also, to avoid doing a bunch of useless FW 4.x stuff on FW 5.x); could you guys double-check that I didn't horribly break it, since I can only test on 4.x over here ;).

beedaddy commented:
> Okay, I pushed a refactor to try to limit code duplication (and, also, to avoid doing a bunch of useless FW 4.x stuff on FW 5.x); could you guys double-check that I didn't horribly break it, since I can only test on 4.x over here ;).

Does not seem to work here.  

[kfmon_dump.log](https://github.com/user-attachments/files/20326635/kfmon_dump.log)

beedaddy commented:
@NiLuJe It fails from 2510b53 on (fwVersion is 0).

beedaddy commented:
I guess that `is_fw_recent_enough()` is only called on devices with AllWinner cpu, whereas the new kobo devices are using a MediaTek cpu. I tried to call this function earlier and now it works (see #16) .

NiLuJe commented:
Ah, of course it turned out to be something stupid, thanks ;).

NiLuJe commented:
Okay, I pushed a few final tweaks, hopefully everything still works ;).

beedaddy commented:
> Okay, I pushed a few final tweaks, hopefully everything still works ;).

Yes, it does. 👍

NiLuJe commented:
Cool, thanks!
